examples/pytorch/language-modeling/run_clm.py

```diff
@@ -513,6 +524,42 @@ def main():
             else:
                 assert len(param.shape) == 2
                 xs.mark_sharding(param, mesh, range(len(param.shape)))
+    elif model_args.spmd_2d_sharding > 0:
+        print('Applying 2D sharding to all parameters')
+        for name, param in model.named_parameters():
+            # Apply 2D sharding:
+            # embedding (model, data)
+            # attn QKV (data, model)
+            # attn O (model, data)
+            # mlp gate, up (data, model)
+            # mlp down (model, data)
+            print('> Sharding tensor', name, param.shape)
+            mod = model_args.spmd_2d_sharding
+            data = num_devices // mod
+            assert mod * data == num_devices
+            data_model_mesh = xs.Mesh(device_ids, (data, mod))
+            model_data_mesh = xs.Mesh(device_ids, (mod, data))
+
+            # We don't care about layernorm's weights, and
+            # LLaMA doesn't use biases.
+            if len(param.shape) == 1:
+                continue
+
+            if 'embed_tokens' in name:
+                xs.mark_sharding(param, model_data_mesh, range(len(param.shape)))
+            elif 'q_proj' in name or 'k_proj' in name or 'v_proj' in name:
+                xs.mark_sharding(param, data_model_mesh, range(len(param.shape)))
+            elif 'o_proj' in name:
+                xs.mark_sharding(param, model_data_mesh, range(len(param.shape)))
+            elif 'gate_proj' in name or 'up_proj' in name:
+                xs.mark_sharding(param, data_model_mesh, range(len(param.shape)))
+            elif 'down_proj' in name:
+                xs.mark_sharding(param, model_data_mesh, range(len(param.shape)))
```

Comment:
Just for my understanding: I noticed that [HF](https://github.com/huggingface/transformers/blob/1689aea73346816b936b84932e12b774974e61a6/src/transformers/models/llama/modeling_llama.py#L205-L207) shards `gate_proj` and `up_proj` on the 0th dim and `down_proj` on the 1st dim, but here you're sharding `gate` and `up` on the `data_model` mesh, which places the `model` axis on dim 1.

Is this just a difference in 1- and 2-D sharding?

Reply:
That's a good catch. I don't know. Let me dig into it. I'm following the slides attached on the top of the spreadsheet.

Reply:
No worries! I was just curious, using the sharding from the slides makes sense.

Reply:
Yea, you are right. I have corrected the error.